File: src/csharp/Core/Face.cs
using System.Collections.Generic;
using System.Numerics;
using System.Linq;

namespace DDGCompanion.Core
{
    public class Face
    {
        public HalfEdge? HalfEdge { get; set; }
        public int Index { get; set; } = -1;
        
        public List<Vertex> Vertices()
        {
            var verts = new List<Vertex>();
            if (HalfEdge == null) return verts;
            var he = HalfEdge;
            do
            {
                verts.Add(he.Vertex!);
                he = he.Next!;
            } while (he != HalfEdge);
            return verts;
        }
        
        public double Area()
        {
            var verts = Vertices();
            if (verts.Count != 3) return 0;
            var e1 = verts[1].Position - verts[0].Position;
            var e2 = verts[2].Position - verts[0].Position;
            return 0.5f * Vector3.Cross(e1, e2).Length();
        }
        
        public Vector3 Normal()
        {
            var verts = Vertices();
            if (verts.Count != 3) return Vector3.Zero;
            var e1 = verts[1].Position - verts[0].Position;
            var e2 = verts[2].Position - verts[0].Position;
            return Vector3.Normalize(Vector3.Cross(e1, e2));
        }
    }
}

File: src/cpp/core/Face.cpp
#include "Face.h"
#include "HalfEdge.h"
#include "Vertex.h"

std::vector<Vertex*> Face::vertices() const {
    std::vector<Vertex*> verts;
    if (!halfedge) return verts;
    HalfEdge* he = halfedge;
    do {
        verts.push_back(he->vertex);
        he = he->next;
    } while (he != halfedge);
    return verts;
}

std::vector<HalfEdge*> Face::halfedges() const {
    std::vector<HalfEdge*> hes;
    if (!halfedge) return hes;
    HalfEdge* he = halfedge;
    do {
        hes.push_back(he);
        he = he->next;
    } while (he != halfedge);
    return hes;
}

double Face::area() const {
    if (!halfedge || !isTriangle()) return 0.0;
    auto verts = vertices();
    Eigen::Vector3d e1 = verts[1]->position - verts[0]->position;
    Eigen::Vector3d e2 = verts[2]->position - verts[0]->position;
    return 0.5 * e1.cross(e2).norm();
}

Eigen::Vector3d Face::normal() const {
    if (!halfedge || !isTriangle()) return Eigen::Vector3d::Zero();
    auto verts = vertices();
    Eigen::Vector3d e1 = verts[1]->position - verts[0]->position;
    Eigen::Vector3d e2 = verts[2]->position - verts[0]->position;
    return e1.cross(e2).normalized();
}

bool Face::isTriangle() const {
    if (!halfedge) return false;
    return halfedge->next->next->next == halfedge;
}
